### DISCUSSION TITLE
Signing Air Application without a .pfx file

### DISCUSSION THREAD

**Discussion by Khumee** [airsdk/Adobe-Runtime-Support#3108] [General]
We have a legacy Air Application that we used to sign with DigiCert Certificate using .pfx file. Now certificate authorities don't allow to export .pfx file. Anyone is experiencing the same issue ? . 
Now application can be signed using Digicert key locker or using HSB device but they do not support .airi file extensions.

**Comment by ajwfrost** [comment]
We've got things set up with a remote codesign service (DigiCert ONE) using the below steps, and I believe the same could also be applied to set up signing for using a hardware key...

The main thing I think is that you need a properties file with details of the platform-specific signing provider, see the first steps from:
https://knowledge.digicert.com/tutorials/sign-java-jar-files-with-a-hardware-token-based-code-signing-certificate-in-windows
i.e. you need a configuration file (it can go anywhere!) with a "name" entry to identify this provider, and a library which references the DLL that handles the signing. Note that they've got single backslashes in their example, we've got double backslashes in our file, so maybe see which works (personally I think this is read by Java so should be double-backslash, or single-forwardslash..)

Then you need to let Java know about this .. which you can do by editing your java.security file (normally in JDK_folder/conf/security) and finding where there's a list of "security.provider.nn" entries, and adding one, using the next number in the list:
```
security.provider.14=SunPKCS11 c:/somewhere/where/you/put/that/properties.cfg
```

Then that should all be sorted... and the arguments to AIR/ADT would then be:
```
-storetype pkcs11
-providerName SunPKCS11-nameEntryFromPropertiesFile
-alias Whatever_your_certificate_alias_is
-storepass xxx
```
i.e. the provider name is "SunPKCS11" (same as in the java.security file) hyphenated with the "name" entry from your properties file where you also identified the library.

Hope that helps... it's a bit convoluted to set up, but then should work a charm..
We have this working with the online service where we don't actually need a password because the authentication is done separately, but we still need a 'storepass' argument, hence just pass in 'a' or anything random. For a hardware key I suspect you would still need a password? (although we also use a hardware token for signing our .exe files, and that requires the password to be entered in a pop-up window.. I've never come across a way to pass in the password from the command line!)

Let us know how it goes! If you are using DigiCert and run into any difficulties, let me know and I can try to reproduce the set-up here and then rope in their support if needed.

thanks

**Reply by Khumee** [reply]
Thanks a lot Andrew, I have configured Digicert Signtool and it works with .exe file , In pkcs11.cfg I have 
name=signingmanager
library=C:\Program Files (x86)\Adobe\Adobe Flash Builder 4.6\sdks\3.6.0\lib\adt.jar
slotListIndex=0

https://www.loom.com/share/270dc408a42b4bd0819d23abbb200901

I tried with changing the library to 
library=C:\Program Files (x86)\Adobe\Adobe Flash Builder 4.6\sdks\3.6.0\lib\adt.jar

but it could not sign .airi file 

https://www.loom.com/share/34cdac46b94a4d4c9679757cfba5e2a1

Apparently Digicert support 
https://docs.digicert.com/nl/software-trust-manager/sign-with-digicert-signing-tools/files-supported-for-signing.html

**Reply by ajwfrost** [reply]
So, the process we use in ADT is very similar to how JAR files are signed... if you look under that support list, there's a link for jarsigner:
https://docs.digicert.com/nl/software-trust-manager/sign-with-digicert-signing-tools/third-party-signing-tool-integrations/jarsigner.html
and this has the details for the library and config file:

> PKCS11 library
> Jarsigner uses a configuration file to integrate with Software Trust Manager PKCS11 library.

> Follow these instructions to download [Software Trust Manager PKCS11 library](https://docs.digicert.com/nl/software-trust-manager/tools/cryptographic-libraries-and-frameworks/pkcs11-library.html) and create the configuration file.

You can see the config file under that link, it's a DLL that they've got referenced as the library, and they've saved the file as pkcs11properties.cfg.

It may not help that all their examples have have arguments such as `-providerClass sun.security.pkcs11.SunPKCS11 -providerArg pkcs11properties.cfg` e.g. see [this link](https://docs.digicert.com/nl/software-trust-manager/signing-tools/sign-java-files-with-jarsigner-using-pkcs11-integration.html), but if you set up the java security configuration file, you can replace those two by the `-providerName` that you have in the properties/cfg file (i.e. `signingmanager` in their example).

We perhaps need to try getting the steps better illustrated and working through an example, similar to how their documentation is set up! But hope that helps a bit..

thanks

**Comment by Khumee** [comment]
Hi Andrew,

I configured things for JarSigner and was able to successfully sign a jar file but unfortunately was not able to sign the desired .airi file.

https://www.loom.com/share/1f138f5e41ce40ad8ece967ea8189cbf

I shall be very grateful if you could sign the .airi file available at { https://drive.google.com/file/d/1uAXweFXs1fMsBnP6Y76nXmSkNS1wI75T/view?usp=sharing } in your setup to see if it works or not.Its a kind of show stopper for us to create setup for updates for our clients

**Reply by ajwfrost** [reply]
Hi - I can take a look, but I can't access file sharing systems so can you please upload the files here:
https://transfer.harman.com/requests/E5Sm9gsbtmpsze1TdcF0Ax
thanks

**Comment by Khumee** [comment]
Uploaded file on given link.

**Reply by ajwfrost** [reply]
Thanks .. worked for me with the online DigiCert ONE service. I can try tomorrow with a hardware USB token...

Currently then:
Command line:
```
C:\Program Files\Java\jdk-11\bin>java -jar i:\SDKs\50.2.3.1\lib\adt.jar  -sign -storetype pkcs11 -providerName SunPKCS11-signingmanager -alias xxx_our_keypair_aliasname_xxx -storepass ignored -target air your_file.airi your_file.air
```

I have the file:
```
C:\Program Files\Java\jdk-11\conf\security\java.security
```
with a line:
```
security.provider.14=SunPKCS11 C:/Program Files/DigiCert/DigiCert One Signing Manager Tools/pkcs11properties.cfg
```
and in that `pkcs11properties.cfg` file, I have:
```
name=signingmanager
library="C:\\Program Files\\DigiCert\\DigiCert One Signing Manager Tools\\smpkcs11.dll"
slotListIndex=0
```

Result when I run the .air file:
![image](https://github.com/airsdk/Adobe-Runtime-Support/assets/6400804/f796b275-85d8-43b3-aae3-9add4cb0007f)

**Reply by Khumee** [reply]
Thats great,  we have ray of hope. 

We also have  online DigiCert ONE service and I tried with the command you shared and it asks for the password. Probably it should be ignored and if it is required  Where can i get this password. ?

**Reply by ajwfrost** [reply]
From what I understand, it doesn't actually need a password, so you can just enter anything ("ignored" or just "a", which is used internally by ADT in some cases like this but sadly not in this case!)
I think you also need to check your command-line, the `-providerName` argument needs to be a single value of `SunPKCS11` and whatever the "name" value is in your properties/config file, combined with a hyphen. So for us, `SunPKCS11-signingmanager`, no spaces inbetween those terms.

thanks

**Reply by Khumee** [reply]
Thank Andrew, 

Looks like it is failing for me on last step 
![image](https://github.com/airsdk/Adobe-Runtime-Support/assets/18255710/52283a04-84e0-4e13-a2fc-73b8133abc46)

**Reply by ajwfrost** [reply]
Okay, so we've managed to get our AIRI file signed using a token-based certificate.

The configuration file for it needed to reference a different library:
```
name=signingmanager
library="c:\\windows\\system32\\eTPKCS11.dll"
```

And then for the ADT command line:
- `-storetype` needed to be `pkcs11`
- `-alias` was the full certificate name (common name of the subject of the certificate) - put in quotes because we have spaces and punctuation
- `-storepass` was actually the password that we use to unlock the hardware token for signing.. I think this can be configured so that it's only needed once per user session, but we passed it in (again with quote marks around because of the odd chars it contained)

This worked, we got a .air file that contains the signatures XML file.

And when I ran the .air file, it came up with the "Publisher Identity: VERIFIED" status. (Note that if you already had the application installed, then an installer with the new certificate won't work, you have to do the 'migrate' thing...)

Hope that helps.

**Reply by Khumee** [reply]
Can you please guide on solution for this exception : "unable to create signature exception" 
![image](https://github.com/airsdk/Adobe-Runtime-Support/assets/18255710/1de16013-ec1d-4f6b-aa2a-9bcb1f1c7395)

**Reply by ajwfrost** [reply]
That's where it's trying to create/write the signatures.xml file which is where the signature would be generated; sadly that is a bit of code that catches an exception only to then throw a generic IOException with a standard message that you see there.
So, the actual information for the original exception isn't coming out there.. so it could be for a number of things.

Interesting that it's got that far whilst you still have the "ignored" password, I had to start putting in the proper password for it..

We can maybe look to add more informative output for this sort of condition...

**Reply by Khumee** [reply]
I found following errors in 

time="2024-03-14T18:57:38+05:00" level="error" msg="hash signing failed for hash: 1781865979da924eb84c9de62887e145fc68f51c, keypair_id: 81c01e9d-c9f7-49d6-9cdf-634f3c3a4414, signature_algorithm: SHA1withRSA: status_code=400, message={\"error\":{\"status\":\"invalid_algorithm\",\"message\":\"Signing request failed for keypair 81c01e9d-c9f7-49d6-9cdf-634f3c3a4414. Signature requests for Signature Algorithm SHA1withRSA are not permitted for a keypair associated with a public issued certificate as per compliance guidelines.\"}}, nested_error=<nil>" executable="C:\\\\Program Files\\\\Java\\\\jdk-11\\\\bin\\\\java.exe" func="securesigning/cli/pkcs11/service.(*service).Sign:622" pid="18912"


time="2024-03-14T18:57:38+05:00" level="error" msg="failed to sign, nested_error=\"hash signing failed for hash: 1781865979da924eb84c9de62887e145fc68f51c, keypair_id: 81c01e9d-c9f7-49d6-9cdf-634f3c3a4414, signature_algorithm: SHA1withRSA: status_code=400, message={\"error\":{\"status\":\"invalid_algorithm\",\"message\":\"Signing request failed for keypair 81c01e9d-c9f7-49d6-9cdf-634f3c3a4414. Signature requests for Signature Algorithm SHA1withRSA are not permitted for a keypair associated with a public issued certificate as per compliance guidelines.\"}}, nested_error=<nil>\"" executable="C:\\\\Program Files\\\\Java\\\\jdk-11\\\\bin\\\\java.exe" func="securesigning/cli/pkcs11.(*Context).SignFinal:411" pid="18912

**Reply by ajwfrost** [reply]
I think this may be related to the security settings in your JDK ... are you able to open the `java.security` file (in `c:\program files\java\jdk-11\conf\security`) and look for settings such as `jdk.disabled.namedCurves`, `jdk.jar.disabledAlgorithms` and `jdk.certpath.disabledAlgorithms`?

**Reply by Khumee** [reply]
Yes, I can see the settings, should I comment them ? if possible. can you share yours settings for the above attributes

**Reply by ajwfrost** [reply]
I think just check that they don't have restrictions for RSA  .. it may also depend on your key size but I would have thought if you're using a hardware token, it should be okay..?
```
jdk.certpath.disabledAlgorithms=MD2, MD5, SHA1 jdkCA & usage TLSServer, \
    RSA keySize < 1024, DSA keySize < 1024, EC keySize < 224

jdk.jar.disabledAlgorithms=MD2, MD5, RSA keySize < 1024, \
      DSA keySize < 1024

#jdk.disabled.namedCurves=
```
i.e. I don't have any of the namedCurves things disabled. I don't really know what all this means though, am just suspicious of this kind of setting given the error messages that you pasted in above..!

thanks

**Reply by Khumee** [reply]
Thanks again Andrew, without your support Its difficult to get it here. 

In the error description, it says that "invalid_algorithm". I am using this 
![image](https://github.com/airsdk/Adobe-Runtime-Support/assets/18255710/3d42faa8-2b38-4f44-a679-e76420741c60)

Can you please share the Algorithm you are using with online signing .

**Reply by Khumee** [reply]
H @ajwfrost 

Can you please share your Algorithm details that you used to sign from Digicert One ?

**Reply by ajwfrost** [reply]
Hi

It's ADT which is choosing the algorithm, for signing an .air file it will use `SHA1withRSA`

There may be issues with this algorithm being considered outdated, although I would hope it's something that can still be configured... but we can also look into whether we can cope with other algorithms as part of this process.

I suspect it's Java that's imposing the restriction here, rather than it being related to DigiCert directly. 
e.g. see https://www.java.com/en/configure_crypto.html and the section headed "Disable SHA-1 signed jars" .. although it looks like you've been using JDK 11 and 15 above? But make sure there's nothing like "SHA1 usage SignedJAR & denyAfter 2019-01-01" in your Java security config!

thanks

**Reply by Khumee** [reply]
Thanks again Andrew 
I removed  "SHA1 usage SignedJAR & denyAfter 2019-01-01" but no luck yet 

Can you please share the JDK version you used to sign and also share the java.security file, I will replicate the exact environment as yours

**Reply by ajwfrost** [reply]
I'm using Oracle's JDK 11.0.18, from December 2022:
```
java version "11.0.18" 2023-01-17 LTS
Java(TM) SE Runtime Environment 18.9 (build 11.0.18+9-LTS-195)
Java HotSpot(TM) 64-Bit Server VM 18.9 (build 11.0.18+9-LTS-195, mixed mode)
```
[java.security.txt](https://github.com/airsdk/Adobe-Runtime-Support/files/14636206/java.security.txt)

**Reply by Khumee** [reply]
I tried with exact same version but no luck, I do not think its on JDK side

As I got this response from Digicert 

**Based on the command : 
"C:\Program Files\Java\jdk-11\bin\java" -jar "C:\Program Files\AirSDK\lib\adt.jar"  -sign -storetype pkcs11 -providerName SunPKCS11-signingmanager -alias key_6144*** -storepass ignored -target air D:\UIS.1.23.6530.airi D:\UIS.1.23.6530.air   

it appears that the -sigalg  for signing algorithm is not specified which may have resulted in an attempt to sign using SHA1.

SHA1 is no longer a secure signing method and is not supported by DigiCert Keylocker.**

Still wondering how it works on your end.

Now they routed me again to try with jarsiger with this command 

jarsigner -keystore NONE -storepass NONE -storetype PKCS11 -sigalg SHA256withRSA -providerClass sun.security.pkcs11.SunPKCS11 -providerArg "pkcs11properties.cfg" -signedjar D:\\UIS.1.23.6530.air  D:\\UIS.1.23.6530.airi key_614426073 -tsa "http://timestamp.digicert.com" -verbose
requesting a signature timestamp

and it is giving me this response 

![image](https://github.com/airsdk/Adobe-Runtime-Support/assets/18255710/40e74e8b-9512-4f30-a020-0ee2d2bdb1e7)

Any suggestion you have this on above exception ?

**Reply by ajwfrost** [reply]
Hi

The AIR file isn't quite the same as a JAR file, so I don't think it's possible to use jarsigner for it. But their explanation does make sense and may be the reason for the problem - although yes, I also don't know why it then works for me!!

We can do some checks here on the SHA256withRSA option...

thanks

**Reply by Khumee** [reply]
Can this be considered as feature request to add option of SHA256withRSA as algorithm in adt ?

**Reply by ajwfrost** [reply]
Hi @Khumee - are you able to try this again using the below version of ADT please? from what I can tell, if we do switch the algorithm then the existing runtime app installers are still able to validate the signatures, but it would be good to have that confirmed in the field! So - (a) does the signing of the .air file work, and (b) can you then run that .air file on a computer and see your certificate's common name in the 'Verified publisher' display in the AIR app installer?

thanks
[adt.zip](https://github.com/airsdk/Adobe-Runtime-Support/files/14720654/adt.zip)

**Reply by Khumee** [reply]
Thank you Andrew,,

Let me try it, I will update you in a while

**Reply by Khumee** [reply]
Thanks a million, it worked I can generate a signed air file. ....

https://www.loom.com/share/e648c92c50b94382879f2e94cf9855b5?from_recorder=1&focus_title=1

**Reply by ajwfrost** [reply]
Great! and it shows up okay for the installation?
We can put this into the next update releases (not today's unfortunately though!)

**Reply by Khumee** [reply]
Yes, I am able to install it, 

Just a suggestion to add algorithm as an argument like digicert for jarsigner  
![image](https://github.com/airsdk/Adobe-Runtime-Support/assets/18255710/693c0f87-26ad-4bef-bbf5-50b33eaf8830)

**Reply by ajwfrost** [reply]
We could do that via an argument .. would probably go with defaulting this to sha256 though now, as this is going to be a recurring issue I think! There are different signing processes for the different output files, I think it's only AIR files (and maybe ANEs?) that use this mechanism that had been defaulting to sha1...

**Reply by Khumee** [reply]
@ajwfrost One thing I missed that installer is showing me unknown publisher, 

With Digital key locker I assume that it should be verified or i need to anything as configuration ?

**Reply by ajwfrost** [reply]
So, that part means that the runtime hasn't been able to validate the signature.. which would generally mean, no root of trust (so e.g. like a self-signed certificate which just has the issuer being the same thing.. or a test certificate that may be created from a special/custom/temporary certificate authority).

If your signing certificate should have an issuer trail that goes back to one of the standard roots - like the main DigiCert one, or similar - essentially the same as are used on different operating systems/browsers/etc - then the AIR app installer should work that out and display the appropriate "verified" message.

So if you think it should show up as trusted .. is it something you could share with us and we can see what's happening under the hood within the application validation code?

thanks

**Reply by Khumee** [reply]
Thank you @ajwfrost Please share the link to upload the installer file

**Reply by ajwfrost** [reply]
@Khumee  please use https://transfer.harman.com/requests/bYNDD1gOta2UdINFkUVk1q
thanks

**Reply by Khumee** [reply]
Sent

**Reply by ajwfrost** [reply]
Okay thanks .. this may be something to do with the fact you've got a migration options in there.. 

We'll have to look further into that part; in the meantime, are you able to try it as if it's a new application and sign it just with the new certificate? You'd get an error when running the new .air file, unless you uninstall the old application first (so, feel free to just send us the new .air file so we can try it on a fresh machine here, in case you don't have that set up and don't want to uninstall the older one!)

thanks

**Reply by Khumee** [reply]
I tried without migration but its still the same 
![image](https://github.com/airsdk/Adobe-Runtime-Support/assets/18255710/21b1fe5b-d65d-47c8-9e15-5282ace6f066)

Please share the link for sharing this installer

**Reply by ajwfrost** [reply]
Thanks .. seems to be actually a different mechanism/certificate type being used which may be the problem. But it would definitely help to have this installer please, to keep it simple initially. We've found the migration signature is still using SHA-1...

https://transfer.harman.com/requests/s51iXlahuRIFLvzOi25X0C
thanks

**Reply by Khumee** [reply]
Installer uploaded, 

BTW , i tried it on fresh machine, but same Unknown publisher ..

**Reply by Khumee** [reply]
@ajwfrost Did you get a chance to check the installer without migration ?

**Reply by ajwfrost** [reply]
We did .. and it's also failing; we've narrowed this down to the verification of the signature, but can't actually see where within the code it's finding the problem. The signature validation code is all fairly complex and is from some third party (within Adobe) codebase so it's proving a little difficult to work out. We're trying to look at the comparisons between the signatures from when we do this vs when you do it.... a bit of a painstaking process though!

So, sorry, working on it but it's quite slow..

**Reply by Khumee** [reply]
Hi @ajwfrost , hope you are doing great

This command suddenly stop working 
 java -jar "D:\PLM\Signing\AIRSDK\lib\adt.jar" -sign -storetype pkcs11 -providerName SunPKCS11-signingmanager -alias key_614426073 -storepass ignored -target air UIS.1.23.6537.airi UIS.1.23.6537.air
 
 with error 
 Could not generate timestamp: Connection reset
 
 I tried with  -tsa http://timestamp.digicert.com but it didnot work
 
 Can you please share which timestamp URL is being used by adt internally and how can I override it ?

Regards
Khurram

**Reply by ajwfrost** [reply]
That `-tsa` argument should work.. are you able to check the troubleshooting output to see what the full command-line is being interpreted as? Plus we have some more details (and a workaround using the 'hosts' file) available under the thread for #3419 

We're about to issue a release that has the update for this, internally it would be defaulting to the old Symantec address I believe.

thanks

**Reply by Khumee** [reply]
Thanks @ajwfrost 

-tsa worked now when I inserted after "-alias" argument

**Comment by grvenkataram** [comment]
Can we Sign the Json files with Digicert keylocker provisioning method. I am able to sign in the .dll files with Signtool

**Reply by ajwfrost** [reply]
Not sure about JSON files? But AIR files, yes...
ADT also now accepts arguments more closely aligned to those used by Jarsigner, so that any examples you might find from DigiCert/etc on how to sign a JAR file should also work with ADT (e.g. `-providerClass sun.security.pkcs11.SunPKCS11 -providerArg pkcs11properties.cfg`)

thanks

**Comment by Khumee** [comment]
Hi @ajwfrost 

I am getting this error "Timestamp response not valid" on the below signing command, It was working earlier but somehow stop working now Any idea ???

java -jar "C:\Program Files\AirSDK\lib\adt.jar" -sign -storetype pkcs11 -providerName SunPKCS11-signingmanager -alias key_******* -tsa http://timestamp.digicert.com -storepass ignored -target air UIS.1.23.6540.airi UIS.1.23.6540.air

**Reply by ajwfrost** [reply]
That may depend on the version of AIR SDK you're using: Digicert changed their support for timestamping which we only found out about after the fact.. so you'd need AIR SDK 51.2.1.2 (or ideally .4) to fix this, or you could change the configuration to use an alternative timestamping service such as:
* http://zeitstempel.dfn.de
* http://timestamp.apple.com/ts01

See https://github.com/airsdk/Adobe-Runtime-Support/issues/3767

thanks

**Reply by Khumee** [reply]
@ajwfrost http://zeitstempel.dfn.de/ worked,  thanks a lot. You are always very helpful
